File: javascript/phone-number/phone-number.js
class PhoneNumber {
  constructor(rawData) {
    this.cleanNumber = rawData.replace(/\D/g, '');
  }

  number() {
    if (this.cleanNumber[0] === '1' && this.cleanNumber.length === 11) {
      return this.cleanNumber.slice(1);
    } else if (this.cleanNumber.length === 10) {
      return this.cleanNumber;
    } else {
      return '0000000000';
    }
  }

  areaCode() {
    if (this.number() !== '0000000000') {
      return this.number().slice(0, 3);
    }
  }

  toString() {
    if (this.areaCode()) {
      const firstThreeDigits = this.number().slice(3, 6);
      const lastFourDigits = this.number().slice(6);
      return `(${this.areaCode()}) ${firstThreeDigits}-${lastFourDigits}`;
    }
  }
}

module.exports = PhoneNumber;

File: ruby/phone-number/phone_number.rb
class PhoneNumber
  VALID_LENGTH = 10
  COUNTRY_CODE = '1'
  INVALID_RETURN = '0000000000'

  def initialize(phone_number)
    @phone_number = phone_number
  end

  def number
    @number ||= if clean_number.length == VALID_LENGTH
                  clean_number
                else
                  INVALID_RETURN
                end
  end

  def area_code
    number[0..2]
  end

  def to_s
    first_three = number[3..5]
    last_four = number[6..10]
    "(#{area_code}) #{first_three}-#{last_four}"
  end

  private

  attr_reader :phone_number

  def digits_only
    return INVALID_RETURN if phone_number.match(/[A-z]/)
    phone_number.gsub(/[^0-9]/, '')
  end

  def handle_country_code(digits)
    if digits.length > VALID_LENGTH && digits.chars.first == COUNTRY_CODE
      digits[1..-1]
    else
      digits
    end
  end

  def clean_number
    handle_country_code(digits_only)
  end
end
